lib/report.ts

```diff
@@ -0,0 +1,256 @@
+// node_modules
+import _ from "lodash";
+// lib
+import QueryString from "./query-string";
+import { splitPathAndQueryString } from "./query-utils";
+// types
+import type {
+  Profiles,
+  ProfilesGeneric,
+  ProfilesProps,
+  Schema,
+  SchemaProperties,
+  SchemaPropertiesGenerics,
+  SchemaPropertiesProps,
+  SearchResults,
+  SearchResultsColumns,
+} from "../globals.d";
+
+/**
+ * The report code often references the `ColumnSpec` object that specifies the columns of a
+ * report and only gets used internally to this file.
+ */
+interface ColumnSpec {
+  // Column ID, which is the property name of the column in the report data; e.g. `@id`
+  id: string;
+  // Column title, which is the human-readable name of the column; e.g. "ID"
+  title: string;
+}
+
+/**
+ * Get the schemas for the given types in `reportTypes`. Abstract types within `reportTypes` don't
+ * have their own schemas, but all their subtypes' schemas get included in the result. If a type in
+ * `reportType` doesn't have a schema nor subtypes, it gets ignored. If `profiles` isn't available,
+ * null gets returned.
+ * @param {string[]} reportTypes All `type` in the query string of the report
+ * @param {Profiles|null} profiles All schemas for all types keyed by their `@type`
+ * @returns {Schema[]} Schemas for the given `reportTypes`
+ */
+export function getSchemasForReportTypes(
+  reportTypes: string[],
+  profiles: Profiles | null
+): Schema[] {
+  if (profiles) {
```

Comment:
I avoid early return more than most, probably because of my early training where early return was frowned upon, as it’s kind of a form of goto, which was also frowned upon, and still is. So basically, if the situation feels convenient to avoid early return, I avoid it. Plus if you want to console.log data at the end of a function, you have to search through the function for early returns that also need a console.log, which I find a little painful. I’d like to avoid all early returns, but I know I could never get away with it these days.